### ISSUE TITLE
How to use Flickity Events?

### ISSUE THREAD

**Issue by siberiadev** [drewjbartlett/vue-flickity#45]
Could you tell me how to use events API of Flickity?
```
this.$refs.slider.on('select', function () {
      console.log(vm.$refs.slider.flickity.selectedIndex);
});
```
This code doesnt work.
I get the error "Cannot read property 'on' of undefined"

**Comment by nikwins** [comment]
Try
`this.$refs.flickity.on('select', function () {
            console.log(vm.$refs.flickity.selectedIndex());
        });`

Otherwise just console.log this.$refs to see how your component is named

**Comment by lukepearce** [comment]
If I wrap the event listener in a timeout I don't get the error.

```
setTimeout(() => {
  this.$refs.flickity.on('change', (event) => {
    this.$emit('change', event);
  });
}, 1000);
```

**Comment by siberiadev** [comment]
Yeah, when I console my refs I can see flickity object, but when I’m trying
to call .on event it callback an error I told you before. Any ideas?

ср, 21 нояб. 2018 г. в 17:03, nikwins <notifications@github.com>:

**Comment by siberiadev** [comment]
> setTimeout(() => { this.$refs.flickity.on('change', (event) => { this.$emit('change', event); }); }, 1000);

YEAH!!! IT WORKS!! THANKS

**Comment by OriginalEXE** [comment]
Use the `mounted` hook to bind your events, and put it inside the `this.$nextTick(() => { });`, no need for the setTimeout with a random delay.

**Comment by lukepearce** [comment]
> Use the `mounted` hook to bind your events, and put it inside the `this.$nextTick(() => { });`, no need for the setTimeout with a random delay.

I get `TypeError: Cannot read property 'on' of undefined` when I place in `mounted` using `nextTick`. If I console log `this.$refs.flickity` I can see the component.

**Comment by OriginalEXE** [comment]
That's weird. Could you create a reproduction on codesandbox? I would be interested in finding out what's causing it.

You are immediately rendering it inside a component, right?

**Comment by lukepearce** [comment]
After starting to set this up on codesandbox I remembered that I adapted `vue-flickity` into my own component (to add a few things) and in doing so I'm actually using https://github.com/metafizzy/flickity-bg-lazyload instead of the standard flickity lib. 

I expect it's got something to do with a feature I added that allows me to delay the start of a carousel. Regardless of what it might be, it isn't on you to help me debug that ;)

**Comment by Yakoot** [comment]
> > Use the `mounted` hook to bind your events, and put it inside the `this.$nextTick(() => { });`, no need for the setTimeout with a random delay.
> 
> I get `TypeError: Cannot read property 'on' of undefined` when I place in `mounted` using `nextTick`. If I console log `this.$refs.flickity` I can see the component.

```
<flickity ref="flickity"
                @init="onInit"...
```
and then you can bind your events in onInit method

**Comment by CodyEddings** [comment]
Is there a sample solution for event listening with Vue 3 & TypeScript?

**Comment by CodyEddings** [comment]
Solution for Vue 3 & Typescript event listening [here ](https://stackoverflow.com/questions/68231576/flickity-events-vue3-typescript-error-thisrefs-flickity-on-is-not-a-func/68231786?noredirect=1#comment120593378_68231786)
